core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java

```diff
@@ -128,13 +137,109 @@ public Sep31PostTransactionResponse postTransaction(
+
+    AssetInfo reqAsset = assetService.getAsset(request.getAssetCode(), request.getAssetIssuer());
+    BigDecimal reqAmount = decimal(request.getAmount(), reqAsset);
+    BigDecimal fee = decimal(feeResponse.getAmount(), reqAsset);
+    String feeAsset = String.valueOf(feeResponse.getAsset());
+
+    BigDecimal amountIn;
+    String amountInAsset = reqAsset.getCode();
+    BigDecimal amountOut;
+    String amountOutAsset;
+
+    if (withoutQuote) {
+      // Without quote.
+      amountOutAsset = amountInAsset;
+      if (strictSend) {
+        // amount_in = req.amount
+        // amount_out = amount_in - amount fee
+        amountIn = reqAmount;
+        amountOut = amountIn.subtract(fee);
+      } else {
+        // amount_in = req.amount + fee
+        // amount_out = req.amount
+        amountIn = reqAmount.add(fee);
+        amountOut = reqAmount;
+      }
+    } else {
+      // With quote
+      Sep38Quote quote = sep38QuoteStore.findByQuoteId(request.getQuoteId());
+      txn.setQuoteId(request.getQuoteId());
+      amountOutAsset = quote.getBuyAsset();
+      BigDecimal buyAmount = decimal(quote.getBuyAmount(), reqAsset);
+      BigDecimal sellAmount = decimal(quote.getSellAmount(), reqAsset);
+
+      if (feeResponse.getAsset().equals(quote.getSellAsset())) {
+        // Fee is applied before the quote
+        if (strictSend) {
+          // amount_in=req.amount
+          // amount_out=quote.buy_amount
+          amountIn = reqAmount;
+          amountOut = buyAmount;
+          // Validate:
+          //   quote.sell_amount = amount_in - fee
+          if (sellAmount.compareTo(amountIn.subtract(fee)) != 0) {
+            throw new BadRequestException(
+                "Bad quote. quote.sell_amount does not equal to (amount_in - amount_fee)");
+          }
+        } else {
+          // amount_in = req.amount + fee
+          // amount_out = quote.buy_amount
+          amountIn = reqAmount.add(fee);
+          amountOut = buyAmount;
+          // Validate:
+          //   quote.sell_amount = req.amount
+          if (sellAmount.compareTo(reqAmount) != 0) {
+            throw new BadRequestException(
+                "Bad quote. quote.sell_amount does not equal to (amount_in - amount_fee)");
+          }
+        }
+      } else {
+        // Fee is applied after the quote
+        if (strictSend) {
+          // amount_in=req.amount
+          // amount_out=quote.buy_amount-fee
+          amountIn = reqAmount;
+          amountOut = buyAmount.subtract(fee);
+          // Validate:
+          //   quote.sell_amount = req.amount
+          if (sellAmount.compareTo(reqAmount) != 0) {
+            throw new BadRequestException(
+                "Bad quote. quote.sell_amount does not equal to amount_in");
+          }
+        } else {
+          // amount_in=req.amount
+          // amount_out=quote.buy_amount -fee
+          amountIn = sellAmount;
+          amountOut = buyAmount.subtract(fee);
```

Comment:
I think you're missing a validation here, `amount_in=req.amount` is not being validated.

Reply:
I had it and took it out. But I cannot remember the reason why I took it out. Added it back.